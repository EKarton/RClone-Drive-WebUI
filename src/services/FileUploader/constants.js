export const ActionTypes = Object.freeze({
  UPLOAD_FILE: 'UPLOAD_FILE',
  CANCEL_UPLOAD: 'CANCEL_UPLOAD',
});
