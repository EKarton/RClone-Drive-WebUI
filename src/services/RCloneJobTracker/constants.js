export const ActionTypes = Object.freeze({
  TRACK_NEW_JOB: 'TRACK_NEW_JOB',
  UNTRACK_NEW_JOB: 'UNTRACK_NEW_JOB',
});
