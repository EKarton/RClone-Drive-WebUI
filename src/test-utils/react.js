import { FileViewerProvider } from 'contexts/FileViewerStore';
import { RCloneInfoStateProvider } from 'contexts/RCloneInfoStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState, initialRCloneInfoState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  const renderedComponent = render(
    <RCloneInfoStateProvider defaultState={initialRCloneInfoState}>
      <FileViewerProvider defaultState={initialFileViewerState}>
        <Router history={history}>{component}</Router>
      </FileViewerProvider>
    </RCloneInfoStateProvider>
  );

  return {
    ...renderedComponent,
    history,
  };
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
