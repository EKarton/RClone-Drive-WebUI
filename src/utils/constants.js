export const ICON_SIZE = Object.freeze({
  SMALL: 'small',
  MEDIUM: 'medium',
  LARGE: 'large',
});

export const StatusTypes = Object.freeze({
  LOADING: 'loading',
  SUCCESS: 'success',
  ERROR: 'error',
});

export const JobStatus = Object.freeze({
  ONGOING: 'ONGOING',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR',
});

export const JobTypes = Object.freeze({
  UPLOAD_FILE: 'UPLOAD_FILE',
  MOVE_FILE: 'MOVE_FILE',
  MOVE_FOLDER: 'MOVE_FOLDER',
  RENAME_FILE: 'RENAME_FILE',
  RENAME_FOLDER: 'RENAME_FOLDER',
});

export const COLOR_MODE = Object.freeze({
  LIGHT: 'light',
  DARK: 'dark',
});

export const ImageMimeTypes = new Set(['image/jpeg', 'image/png']);
