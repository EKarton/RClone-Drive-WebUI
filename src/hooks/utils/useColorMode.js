import { useContext } from 'react';
import { ColorModeContext } from 'contexts/ColorMode/index';

export default function useColorMode() {
  const context = useContext(ColorModeContext);

  if (context === undefined) {
    throw new Error('useColorMode() needs to be used inside ColorModeProvider');
  }

  return context;
}
