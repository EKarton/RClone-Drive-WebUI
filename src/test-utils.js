import { FileViewerProvider, initialState } from 'contexts/FileViewerStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState = initialState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  return render(
    <FileViewerProvider defaultState={initialFileViewerState}>
      <Router history={history}>{component}</Router>
    </FileViewerProvider>
  );
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
