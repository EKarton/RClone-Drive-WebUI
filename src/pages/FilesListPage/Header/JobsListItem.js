import ListItem from '@mui/material/ListItem';
import ListItemText from '@mui/material/ListItemText';
import { useEffect, useState } from 'react';
import { JobStatus, JobTypes } from 'utils/constants';
import { getFullPath } from 'utils/filename-utils';

export default function JobsListItem({ job }) {
  const [status, setStatus] = useState(null);

  useEffect(() => {
    const subscriber = job.status.subscribe((newStatus) => {
      setStatus(newStatus);
    });

    return () => {
      subscriber.unsubscribe();
    };
  });

  const getMessage = () => {
    switch (job.jobType) {
      case JobTypes.MOVE_FILE:
      case JobTypes.MOVE_FOLDER: {
        return getMoveMessage();
      }
      case JobTypes.RENAME_FILE:
      case JobTypes.RENAME_FOLDER: {
        return getRenameMessage();
      }
      case JobTypes.UPLOAD_FILE: {
        return getUploadMessage();
      }
      default: {
        return getDefaultString();
      }
    }
  };

  const getMoveMessage = () => {
    const { src } = job;
    const srcStr = getFileString(src);

    if (status === JobStatus.ERROR) {
      return `Failed to move ${srcStr}`;
    }

    if (status === JobStatus.SUCCESS) {
      return `Moved ${srcStr}`;
    }

    return `Moving ${srcStr}`;
  };

  const getRenameMessage = () => {
    const { src, target } = job;

    if (status === JobStatus.ERROR) {
      return `Failed to rename ${src.name}`;
    }

    if (status === JobStatus.SUCCESS) {
      return `Renamed ${src.name} to ${target.name}`;
    }

    return `Renaming ${src.name} to ${target.name}`;
  };

  const getUploadMessage = () => {
    const { name } = job;

    if (status === JobStatus.ERROR) {
      return `Failed to upload ${name}`;
    }

    if (status === JobStatus.SUCCESS) {
      return `Uploaded ${name}`;
    }

    return `Uploading ${name}`;
  };

  const getDefaultString = () => {
    if (status === JobStatus.ERROR) {
      return 'Failed job';
    }

    if (status === JobStatus.SUCCESS) {
      return 'Job succeeded';
    }

    return 'Job running';
  };

  const getFileString = (file) => {
    return getFullRemotePath(file.remote, getFullPath(file.dirPath, file.name));
  };

  const getFullRemotePath = (remote, fullPath) => {
    return `${remote}:${fullPath}`;
  };

  return (
    <ListItem>
      <ListItemText primary={getMessage()} />
    </ListItem>
  );
}
