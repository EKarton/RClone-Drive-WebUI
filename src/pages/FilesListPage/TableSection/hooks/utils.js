import { JobStatus, JobTypes } from 'utils/constants';

export function getFileNames(data) {
  return new Set(data.filter((file) => !file.IsDir).map((file) => file.Name));
}

export function getFolderNames(data) {
  return new Set(data.filter((file) => file.IsDir).map((file) => file.Name));
}

export function getUploadingJobs(jobs, remote, path) {
  return jobs.filter(
    (job) =>
      job.jobType === JobTypes.UPLOAD_FILE &&
      job.remote === remote &&
      job.dirPath.startsWith(path) &&
      job.status.value !== JobStatus.SUCCESS
  );
}

export function getUploadingFiles(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath === path);
}

export function getUploadingFolders(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath !== path);
}

export function getMovingJobs(jobs, remote, path) {
  const supportedJobTypes = new Set([
    JobTypes.MOVE_FILE,
    JobTypes.MOVE_FOLDER,
    JobTypes.RENAME_FILE,
    JobTypes.RENAME_FOLDER,
  ]);

  return jobs.filter((job) => {
    const correctJob = supportedJobTypes.has(job.jobType);
    const inProgress = job.status.value === JobStatus.ONGOING;

    if (!correctJob || !inProgress) {
      return false;
    }

    const isMatchSrc = job.src?.remote === remote && job.src?.dirPath === path;
    const isMatchTarget = job.target?.remote === remote && job.target?.dirPath === path;

    return correctJob && inProgress && (isMatchSrc || isMatchTarget);
  });
}
