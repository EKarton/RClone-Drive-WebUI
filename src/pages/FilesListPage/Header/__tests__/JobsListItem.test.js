import { BehaviorSubject } from 'rxjs';
import { JobStatus, JobTypes } from 'utils/constants';
import { render } from 'test-utils/react';
import JobsListItem from '../JobsListItem';

describe('JobsListItem', () => {
  it.each([JobStatus.ONGOING, JobStatus.SUCCESS, JobStatus.ERROR])(
    'should match snapshot given job type is UPLOAD_FILE and current status is %s',
    (status) => {
      const { baseElement } = render(
        <JobsListItem
          job={{
            status: new BehaviorSubject(status),
            jobType: JobTypes.UPLOAD_FILE,
            remote: 'gdrive',
            dirPath: 'Documents',
            name: 'Dog.png',
          }}
        />
      );

      expect(baseElement).toMatchSnapshot();
    }
  );

  it.each([JobStatus.ONGOING, JobStatus.SUCCESS, JobStatus.ERROR])(
    'should match snapshot given job type is MOVE_FILE and current status is %s',
    (status) => {
      const { baseElement } = render(
        <JobsListItem
          job={{
            status: new BehaviorSubject(status),
            jobType: JobTypes.MOVE_FILE,
            src: {
              remote: 'gdrive',
              dirPath: 'Documents',
              name: 'Dog.png',
            },
            target: {
              remote: 'box',
              dirPath: 'Archives',
              name: 'Dog.png',
            },
          }}
        />
      );

      expect(baseElement).toMatchSnapshot();
    }
  );

  it.each([JobStatus.ONGOING, JobStatus.SUCCESS, JobStatus.ERROR])(
    'should match snapshot given job type is RENAME_FILE and current status is %s',
    (status) => {
      const { baseElement } = render(
        <JobsListItem
          job={{
            status: new BehaviorSubject(status),
            jobType: JobTypes.RENAME_FILE,
            src: {
              remote: 'gdrive',
              dirPath: 'Documents',
              name: 'Dog.png',
            },
            target: {
              remote: 'box',
              dirPath: 'Archives',
              name: 'Dog.png',
            },
          }}
        />
      );

      expect(baseElement).toMatchSnapshot();
    }
  );

  it.each([JobStatus.ONGOING, JobStatus.SUCCESS, JobStatus.ERROR])(
    'should match snapshot given unknown job type and current status is %s',
    (status) => {
      const { baseElement } = render(
        <JobsListItem
          job={{
            status: new BehaviorSubject(status),
            jobType: 'Random job type',
          }}
        />
      );

      expect(baseElement).toMatchSnapshot();
    }
  );
});
