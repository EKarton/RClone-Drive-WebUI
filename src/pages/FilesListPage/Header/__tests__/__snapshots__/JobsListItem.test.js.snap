// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobsListItem should match snapshot given job type is MOVE_FILE and current status is ERROR 1`] = `
<body>
  <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Failed to move gdrive:Documents/Dog.png
        </span>
      </div>
    </li>
  </div>
</body>
`;

exports[`JobsListItem should match snapshot given job type is MOVE_FILE and current status is ONGOING 1`] = `
<body>
  <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Moving gdrive:Documents/Dog.png
        </span>
      </div>
    </li>
  </div>
</body>
`;

exports[`JobsListItem should match snapshot given job type is MOVE_FILE and current status is SUCCESS 1`] = `
<body>
  <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Moved gdrive:Documents/Dog.png
        </span>
      </div>
    </li>
  </div>
</body>
`;

exports[`JobsListItem should match snapshot given job type is RENAME_FILE and current status is ERROR 1`] = `
<body>
  <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Failed to rename Dog.png
        </span>
      </div>
    </li>
  </div>
</body>
`;

exports[`JobsListItem should match snapshot given job type is RENAME_FILE and current status is ONGOING 1`] = `
<body>
  <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Renaming Dog.png to Dog.png
        </span>
      </div>
    </li>
  </div>
</body>
`;

exports[`JobsListItem should match snapshot given job type is RENAME_FILE and current status is SUCCESS 1`] = `
<body>
  <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Renamed Dog.png to Dog.png
        </span>
      </div>
    </li>
  </div>
</body>
`;

exports[`JobsListItem should match snapshot given job type is UPLOAD_FILE and current status is ERROR 1`] = `
<body>
  <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Failed to upload Dog.png
        </span>
      </div>
    </li>
  </div>
</body>
`;

exports[`JobsListItem should match snapshot given job type is UPLOAD_FILE and current status is ONGOING 1`] = `
<body>
  <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Uploading Dog.png
        </span>
      </div>
    </li>
  </div>
</body>
`;

exports[`JobsListItem should match snapshot given job type is UPLOAD_FILE and current status is SUCCESS 1`] = `
<body>
  <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Uploaded Dog.png
        </span>
      </div>
    </li>
  </div>
</body>
`;

exports[`JobsListItem should match snapshot given unknown job type and current status is ERROR 1`] = `
<body>
  <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Failed job
        </span>
      </div>
    </li>
  </div>
</body>
`;

exports[`JobsListItem should match snapshot given unknown job type and current status is ONGOING 1`] = `
<body>
  <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Job running
        </span>
      </div>
    </li>
  </div>
</body>
`;

exports[`JobsListItem should match snapshot given unknown job type and current status is SUCCESS 1`] = `
<body>
  <div>
    <li
      class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Job succeeded
        </span>
      </div>
    </li>
  </div>
</body>
`;
